File: src/dantofa/core/digitalocean/clusters.py
"""Application logic for DigitalOcean Kubernetes (DOKS) cluster management.

Framework-free: no typer, no SDK imports. The cluster API is reached through a
structurally-typed :class:`SupportsClusterApi` client (the concrete adapter lives
in ``dantofa.clients.digitalocean.clusters``), which keeps this module trivially
unit-testable with a fake/mock client.

These functions mirror the DigitalOcean API one-to-one (create, update, delete,
list) and do no validation of their own — DigitalOcean is the authority on valid
regions/versions/sizes, and its error responses are surfaced to the user as-is.
"""

from __future__ import annotations

import time
from collections.abc import Callable, Mapping, Sequence
from typing import Protocol, cast

_RUNNING_STATE = "running"
# States from which a freshly-created cluster will never reach "running".
_FAILED_STATES = frozenset({"error", "deleted", "deleting"})
_DEFAULT_WAIT_TIMEOUT = 900.0  # seconds (~15 min)
_DEFAULT_POLL_INTERVAL = 10.0  # seconds


class SupportsClusterApi(Protocol):
    """The cluster-API surface this module depends on (see clients adapter)."""

    def list(self) -> list[dict[str, object]]: ...
    def create(self, body: Mapping[str, object]) -> dict[str, object]: ...
    def update(
        self,
        cluster_id: str,
        body: Mapping[str, object],
    ) -> dict[str, object]: ...
    def delete(self, cluster_id: str) -> None: ...
    def get(self, cluster_id: str) -> dict[str, object]: ...
    def get_kubeconfig(self, cluster_id: str) -> str: ...


class ClusterNotFoundError(LookupError):
    """No cluster matched the given name or id."""

    def __init__(self, identifier: str) -> None:
        super().__init__(f"No cluster found matching {identifier!r}.")


class ClusterNotReadyError(RuntimeError):
    """A cluster did not reach the running state (terminal failure or timeout)."""

    def __init__(self, name: str, state: str, *, timed_out: bool = False) -> None:
        reason = "timed out waiting" if timed_out else f"entered state {state!r}"
        super().__init__(f"Cluster {name!r} did not become ready: {reason}.")


def build_node_pool(
    *,
    name: str,
    size: str,
    count: int,
    min_nodes: int,
    max_nodes: int,
) -> dict[str, object]:
    """Assemble a node pool (opinionated: autoscaling is always on)."""
    return {
        "name": name,
        "size": size,
        "count": count,
        "auto_scale": True,
        "min_nodes": min_nodes,
        "max_nodes": max_nodes,
    }


def build_create_body(
    *,
    name: str,
    region: str,
    version: str,
    node_pool: dict[str, object],
    tags: Sequence[str] = (),
    ha: bool = False,
) -> dict[str, object]:
    """Assemble the create body (POST /v2/kubernetes/clusters).

    Takes the single node pool the CLI ever creates and wraps it in the array the
    API expects. Opinionated: auto-upgrade and surge-upgrade are always enabled.
    """
    body: dict[str, object] = {
        "name": name,
        "region": region,
        "version": version,
        "node_pools": [node_pool],
        "auto_upgrade": True,
        "surge_upgrade": True,
        "tags": list(tags),
    }
    if ha:
        body["ha"] = ha
    return body


def build_update_body(
    *,
    tags: list[str] | None = None,
    ha: bool = False,
) -> dict[str, object]:
    """Assemble the update body (PUT /v2/kubernetes/clusters/{id}).

    Opinionated: auto-upgrade and surge-upgrade are always re-asserted as enabled.
    ``tags`` and ``ha`` are included only when supplied.
    """
    body: dict[str, object] = {
        "auto_upgrade": True,
        "surge_upgrade": True,
    }
    if tags is not None:
        body["tags"] = tags
    if ha:
        body["ha"] = ha
    return body


def list_clusters(client: SupportsClusterApi) -> list[dict[str, object]]:
    return client.list()


def create_cluster(
    client: SupportsClusterApi,
    body: Mapping[str, object],
) -> dict[str, object]:
    """Create a cluster. Returns the cluster DigitalOcean reports back."""
    return client.create(body)


def _resolve(
    clusters: list[dict[str, object]],
    name: str,
) -> dict[str, object] | None:
    """Find a cluster by name. Clusters are identified by name only."""
    return next((c for c in clusters if c.get("name") == name), None)


def update_cluster(
    client: SupportsClusterApi,
    name: str,
    body: Mapping[str, object],
) -> dict[str, object]:
    """Update the named cluster with the given mutable fields.

    Clusters are identified by name (the id is resolved internally for the API
    call). The DO update endpoint requires ``name``, so it is always re-sent;
    there is no rename option.
    """
    existing = _resolve(client.list(), name)
    if existing is None:
        raise ClusterNotFoundError(name)
    payload: dict[str, object] = {"name": name, **body}
    return client.update(str(existing["id"]), payload)


def get_kubeconfig(client: SupportsClusterApi, name: str) -> str:
    """Return the kubeconfig for the named cluster."""
    existing = _resolve(client.list(), name)
    if existing is None:
        raise ClusterNotFoundError(name)
    return client.get_kubeconfig(str(existing["id"]))


def _state(cluster: Mapping[str, object]) -> str | None:
    status = cast("Mapping[str, object]", cluster.get("status") or {})
    state = status.get("state")
    return state if isinstance(state, str) else None


def wait_for_running(
    client: SupportsClusterApi,
    name: str,
    *,
    timeout: float = _DEFAULT_WAIT_TIMEOUT,
    interval: float = _DEFAULT_POLL_INTERVAL,
    sleep: Callable[[float], object] = time.sleep,
    monotonic: Callable[[], float] = time.monotonic,
) -> dict[str, object]:
    """Poll until the named cluster reaches the running state; return it.

    Raises :class:`ClusterNotReadyError` on a terminal failure state or when
    ``timeout`` seconds elapse. ``sleep``/``monotonic`` are injectable for tests.
    """
    existing = _resolve(client.list(), name)
    if existing is None:
        raise ClusterNotFoundError(name)
    cluster_id = str(existing["id"])
    deadline = monotonic() + timeout
    while True:
        # SKY-D216 false positive: get() is a cluster-id Protocol call, not an
        # HTTP/URL sink — the adapter owns the request.
        cluster = client.get(cluster_id)  # skylos: ignore[SKY-D216]
        state = _state(cluster)
        if state == _RUNNING_STATE:
            return cluster
        if state in _FAILED_STATES:
            raise ClusterNotReadyError(name, state or "unknown")
        if monotonic() >= deadline:
            raise ClusterNotReadyError(name, state or "unknown", timed_out=True)
        _ = sleep(interval)


def delete_cluster(client: SupportsClusterApi, name: str) -> str | None:
    """Delete the named cluster. Idempotent: a missing cluster is a no-op.

    Returns the resolved cluster id when one was deleted, or ``None`` if no
    cluster by that name exists.
    """
    match = _resolve(client.list(), name)
    if match is None:
        return None
    cluster_id = str(match["id"])
    client.delete(cluster_id)  # skylos: ignore[SKY-D216]
    return cluster_id

File: internal/core/digitalocean/clusters.go
package digitalocean

import (
	"context"
	"fmt"
	"time"
)

const runningState = "running"

// failedStates are terminal: a freshly-created cluster in one will never reach
// running.
var failedStates = map[string]bool{"error": true, "deleted": true, "deleting": true}

// DefaultWaitTimeout / DefaultPollInterval govern `create --wait`.
const (
	DefaultWaitTimeout  = 15 * time.Minute
	DefaultPollInterval = 10 * time.Second
)

// Cluster is a DOKS cluster as surfaced to the user.
type Cluster struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	Region    string     `json:"region"`
	Version   string     `json:"version"`
	State     string     `json:"state,omitempty"`
	Endpoint  string     `json:"endpoint,omitempty"`
	Tags      []string   `json:"tags,omitempty"`
	NodePools []NodePool `json:"node_pools,omitempty"`
	CreatedAt string     `json:"created_at,omitempty"`
}

// NodePool is a cluster node pool as surfaced to the user.
type NodePool struct {
	ID        string `json:"id,omitempty"`
	Name      string `json:"name"`
	Size      string `json:"size"`
	Count     int    `json:"count"`
	AutoScale bool   `json:"auto_scale"`
	MinNodes  int    `json:"min_nodes"`
	MaxNodes  int    `json:"max_nodes"`
}

// NodePoolSpec describes a node pool to create.
type NodePoolSpec struct {
	Name      string
	Size      string
	Count     int
	MinNodes  int
	MaxNodes  int
	AutoScale bool
}

// CreateSpec is the neutral create request. The opinionated invariants
// (auto-upgrade, surge-upgrade, autoscaling) are baked in by the builders here,
// not by the client adapter, so no adapter can bypass them.
type CreateSpec struct {
	Name         string
	Region       string
	Version      string
	NodePools    []NodePoolSpec
	Tags         []string
	AutoUpgrade  bool
	SurgeUpgrade bool
	HA           bool
}

// UpdateSpec is the neutral update request. Tags is nil to leave tags untouched,
// or a (possibly empty) slice to replace them. Name is always re-sent.
type UpdateSpec struct {
	Name         string
	Tags         *[]string
	AutoUpgrade  bool
	SurgeUpgrade bool
	HA           bool
}

// ClusterAPI is the DO cluster surface this package depends on.
type ClusterAPI interface {
	List(ctx context.Context) ([]Cluster, error)
	Create(ctx context.Context, spec CreateSpec) (Cluster, error)
	Update(ctx context.Context, id string, spec UpdateSpec) (Cluster, error)
	Delete(ctx context.Context, id string) error
	Get(ctx context.Context, id string) (Cluster, error)
	GetKubeconfig(ctx context.Context, id string) (string, error)
}

// ClusterNotFoundError is returned when no cluster matches a name.
type ClusterNotFoundError struct{ Identifier string }

func (e *ClusterNotFoundError) Error() string {
	return fmt.Sprintf("no cluster found matching %q", e.Identifier)
}

// ClusterNotReadyError is returned when a cluster fails to reach running.
type ClusterNotReadyError struct {
	Name     string
	State    string
	TimedOut bool
}

func (e *ClusterNotReadyError) Error() string {
	reason := fmt.Sprintf("entered state %q", e.State)
	if e.TimedOut {
		reason = "timed out waiting"
	}
	return fmt.Sprintf("cluster %q did not become ready: %s", e.Name, reason)
}

// BuildNodePool assembles a node pool spec (opinionated: autoscaling is always on).
func BuildNodePool(name, size string, count, minNodes, maxNodes int) NodePoolSpec {
	return NodePoolSpec{
		Name: name, Size: size, Count: count,
		MinNodes: minNodes, MaxNodes: maxNodes, AutoScale: true,
	}
}

// BuildCreateSpec assembles a create spec around the single node pool the CLI
// creates (opinionated: auto-upgrade and surge-upgrade are always enabled).
func BuildCreateSpec(name, region, version string, pool NodePoolSpec, tags []string, ha bool) CreateSpec {
	if tags == nil {
		tags = []string{}
	}
	return CreateSpec{
		Name: name, Region: region, Version: version,
		NodePools: []NodePoolSpec{pool}, Tags: tags,
		AutoUpgrade: true, SurgeUpgrade: true, HA: ha,
	}
}

// BuildUpdateSpec assembles an update spec (opinionated: auto-upgrade and
// surge-upgrade are always re-asserted).
func BuildUpdateSpec(tags *[]string, ha bool) UpdateSpec {
	return UpdateSpec{Tags: tags, AutoUpgrade: true, SurgeUpgrade: true, HA: ha}
}

// ListClusters returns every cluster.
func ListClusters(ctx context.Context, client ClusterAPI) ([]Cluster, error) {
	return client.List(ctx)
}

// CreateCluster creates a cluster and returns what DigitalOcean reports back.
func CreateCluster(ctx context.Context, client ClusterAPI, spec CreateSpec) (Cluster, error) {
	return client.Create(ctx, spec)
}

func resolve(clusters []Cluster, name string) (Cluster, bool) {
	for _, c := range clusters {
		if c.Name == name {
			return c, true
		}
	}
	return Cluster{}, false
}

// UpdateCluster updates the named cluster's mutable fields. Clusters are
// identified by name; the id is resolved internally.
func UpdateCluster(ctx context.Context, client ClusterAPI, name string, spec UpdateSpec) (Cluster, error) {
	clusters, err := client.List(ctx)
	if err != nil {
		return Cluster{}, err
	}
	existing, ok := resolve(clusters, name)
	if !ok {
		return Cluster{}, &ClusterNotFoundError{Identifier: name}
	}
	spec.Name = name
	return client.Update(ctx, existing.ID, spec)
}

// GetKubeconfig returns the kubeconfig for the named cluster.
func GetKubeconfig(ctx context.Context, client ClusterAPI, name string) (string, error) {
	clusters, err := client.List(ctx)
	if err != nil {
		return "", err
	}
	existing, ok := resolve(clusters, name)
	if !ok {
		return "", &ClusterNotFoundError{Identifier: name}
	}
	return client.GetKubeconfig(ctx, existing.ID)
}

// WaitForRunning polls until the named cluster reaches running, returning it.
// It fails on a terminal state or when timeout elapses.
func WaitForRunning(ctx context.Context, client ClusterAPI, name string, timeout, interval time.Duration) (Cluster, error) {
	clusters, err := client.List(ctx)
	if err != nil {
		return Cluster{}, err
	}
	existing, ok := resolve(clusters, name)
	if !ok {
		return Cluster{}, &ClusterNotFoundError{Identifier: name}
	}
	deadline := time.Now().Add(timeout)
	for {
		cluster, err := client.Get(ctx, existing.ID)
		if err != nil {
			return Cluster{}, err
		}
		switch {
		case cluster.State == runningState:
			return cluster, nil
		case failedStates[cluster.State]:
			return Cluster{}, &ClusterNotReadyError{Name: name, State: cluster.State}
		case time.Now().After(deadline):
			return Cluster{}, &ClusterNotReadyError{Name: name, State: cluster.State, TimedOut: true}
		}
		select {
		case <-ctx.Done():
			return Cluster{}, ctx.Err()
		case <-time.After(interval):
		}
	}
}

// DeleteCluster deletes the named cluster. Idempotent: a missing cluster is a
// no-op. Returns the resolved id, or "" if none existed.
func DeleteCluster(ctx context.Context, client ClusterAPI, name string) (string, error) {
	clusters, err := client.List(ctx)
	if err != nil {
		return "", err
	}
	existing, ok := resolve(clusters, name)
	if !ok {
		return "", nil
	}
	if err := client.Delete(ctx, existing.ID); err != nil {
		return "", err
	}
	return existing.ID, nil
}
